src/streams.rs

```diff
@@ -0,0 +1,669 @@
+        let rows: SRRows = from_redis_value(v)?;
+        let mut reply = StreamReadReply::default();
+        for row in &rows {
+            for (key, entry) in row.iter() {
+                let mut k = StreamKey::default();
+                k.key = key.to_owned();
+                for id_row in entry {
+                    let mut i = StreamId::default();
+                    for (id, map) in id_row.iter() {
+                        i.id = id.to_owned();
+                        i.map = map.to_owned();
+                    }
+                    k.ids.push(i);
+                }
+                reply.keys.push(k);
+            }
+        }
+        Ok(reply)
+    }
+}
+
+impl FromRedisValue for StreamRangeReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let rows: Vec<HashMap<String, HashMap<String, Value>>> = from_redis_value(v)?;
+        let mut reply = StreamRangeReply::default();
+        for row in &rows {
+            let mut i = StreamId::default();
+            for (id, map) in row.iter() {
+                i.id = id.to_owned();
+                i.map = map.to_owned();
+            }
+            reply.ids.push(i);
+        }
+        Ok(reply)
+    }
+}
+
+impl FromRedisValue for StreamClaimReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let rows: Vec<HashMap<String, HashMap<String, Value>>> = from_redis_value(v)?;
+        let mut reply = StreamClaimReply::default();
+        for row in &rows {
+            let mut i = StreamId::default();
+            for (id, map) in row.iter() {
+                i.id = id.to_owned();
+                i.map = map.to_owned();
+            }
+            reply.ids.push(i);
+        }
+        Ok(reply)
+    }
+}
+
+impl FromRedisValue for StreamPendingReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let parts: (usize, Option<String>, Option<String>, Vec<Vec<String>>) = from_redis_value(v)?;
+        let count = parts.0.to_owned() as usize;
+
+        if count == 0 {
+            Ok(StreamPendingReply::Empty)
+        } else {
+            let mut result = StreamPendingData::default();
+
+            let start_id = match parts.1.to_owned() {
+                Some(start) => Ok(start),
+                None => Err(Error::new(
+                    ErrorKind::Other,
+                    "IllegalState: Non-zero pending expects start id",
+                )),
+            }?;
+
+            let end_id = match parts.2.to_owned() {
+                Some(end) => Ok(end),
+                None => Err(Error::new(
+                    ErrorKind::Other,
+                    "IllegalState: Non-zero pending expects end id",
+                )),
+            }?;
+
+            result.count = count;
+            result.start_id = start_id;
+            result.end_id = end_id;
+
+            for consumer in &parts.3 {
+                let mut info = StreamInfoConsumer::default();
+                info.name = consumer[0].to_owned();
```

Comment:
Unncessary to_owned

Reply:
Compile reports that it's behind a shared reference

Reply:
If this is known to just be a two element sequence you can extract it as a tuple instead of a vec which will let you easily move it.

Reply:
Fixed!